src/btree/bt_sync_obsolete.c

```diff
@@ -392,30 +428,41 @@ __checkpoint_cleanup_page_skip(
     }
 
     /*
-     * The checkpoint cleanup fast deletes the obsolete leaf page by marking it as deleted
-     * in the internal page. To achieve this,
+     * From this point, the page is on disk and checkpoint cleanup should NOT induce a read if at
+     * least of one of the following conditions is met:
      *
-     * 1. Checkpoint has to read all the internal pages that have obsolete leaf pages.
-     *    To limit the reading of number of internal pages, the aggregated stop durable timestamp
-     *    is checked except when the table is logged. Logged tables do not use timestamps.
+     * - The page is a leaf page with no overflow items because obsolete leaf pages with overflow
+     * keys/values cannot be fast deleted to free the overflow blocks. The overflow blocks can be
+     * freed during reconciliation after being marked dirty.
      *
-     * 2. Obsolete leaf pages with overflow keys/values cannot be fast deleted to free
-     *    the overflow blocks. Read the page into cache and mark it dirty to remove the
-     *    overflow blocks during reconciliation.
+     * - The page does not have any deletes (checked using the aggregated stop durable timestamp)
+     * AND
+     *  - the checkpoint cleanup is configured to skip internal pages OR
+     *  - the table is not logged (they do not use timestamps).
      *
      * FIXME: Read internal pages from non-logged tables when the remove/truncate
      * operation is performed using no timestamp.
      */
-
-    if (addr.type == WT_ADDR_LEAF_NO ||
-      (addr.ta.newest_stop_durable_ts == WT_TS_NONE &&
-        (F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_SKIP_INT) ||
-          !F_ISSET(S2BT(session), WT_BTREE_LOGGED)))) {
-        __wt_verbose_debug2(
-          session, WT_VERB_CHECKPOINT_CLEANUP, "%p: page walk skipped", (void *)ref);
-        WT_STAT_CONN_DSRC_INCR(session, checkpoint_cleanup_pages_walk_skipped);
-        *skipp = true;
+    skip_intl_pages = F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_SKIP_INT);
+    logged = F_ISSET(S2BT(session), WT_BTREE_LOGGED);
+    page_has_deletes = addr.ta.newest_stop_durable_ts != WT_TS_NONE;
+    if (addr.type == WT_ADDR_LEAF_NO || (!page_has_deletes && (skip_intl_pages || !logged))) {
```

Comment:
I am not sure I understand the logic with `skip_intl_pages`.
Indeed, `addr.type` could be either `WT_ADDR_INT` or `WT_ADDR_LEAF`. Shouldn't we apply this setting only if the type is `WT_ADDR_INT` and we want to skip internal pages?

Reply:
This `skip_intl_pages` logic is introduced to avoid reading internal pages from the disk for a logged table. Reading the internal pages causes performance regression. 

Usually for a MongoDB database, there will not be any overflow pages. Changing this check can allow reading the overflow pages of a logged table which is not required.

I am fine with the existing code.

Reply:
As discussed on Slack, it is being reworked in WT-13333. I will merge this PR with develop once WT-13333 is completed.